File: src/actor.ts
export type Actor = AnonymousActor | MemberActor | SystemActor;

export const SERVICE_BACKEND = "geul-backend";
export const SERVICE_WEB = "geul-web";
export const SERVICE_IDENTITY = "geul-identity";
export const SERVICE_EDITOR_COLLAB = "geul-collab";
export const SERVICE_CDN = "geul-cdn";
export const SERVICE_OG = "geul-og";
export const SERVICE_ASSET_OPTIMIZER = "geul-asset-optimizer";
export const SERVICE_TRANSCODER = "geul-transcoder";
export const SERVICE_WAVEFORM_PROCESSOR = "geul-waveform-processor";

export const CANONICAL_SERVICE_NAMES = [
  SERVICE_BACKEND,
  SERVICE_WEB,
  SERVICE_IDENTITY,
  SERVICE_EDITOR_COLLAB,
  SERVICE_CDN,
  SERVICE_OG,
  SERVICE_ASSET_OPTIMIZER,
  SERVICE_TRANSCODER,
  SERVICE_WAVEFORM_PROCESSOR,
] as const;

export type ServiceName = (typeof CANONICAL_SERVICE_NAMES)[number];

const canonicalServiceNames = new Set<string>(CANONICAL_SERVICE_NAMES);

export function parseServiceName(value: string): ServiceName {
  if (!canonicalServiceNames.has(value)) {
    throw new TypeError(
      `unknown canonical service name ${JSON.stringify(value)}`,
    );
  }
  return value as ServiceName;
}

export function instrumentationName(
  serviceName: ServiceName,
  component: string,
): string {
  return `${serviceName}/${component}`;
}

export interface AnonymousActor {
  readonly kind: "anonymous";
}

export interface MemberActor {
  readonly kind: "member";
  readonly sessionId: string;
  readonly identityId: string;
  readonly memberId: string;
}

export interface SystemActor {
  readonly kind: "system";
  readonly serviceName: ServiceName;
}

export type RecordActor =
  | { readonly actor_kind: "anonymous" }
  | { readonly actor_kind: "member"; readonly actor_member_id: string }
  | { readonly actor_kind: "system"; readonly actor_service: string };

export function actorForRecord(actor: Actor): RecordActor {
  switch (actor.kind) {
    case "anonymous":
      return { actor_kind: "anonymous" };
    case "member":
      if (actor.memberId === "") {
        throw new TypeError("member actor requires memberId");
      }
      return { actor_kind: "member", actor_member_id: actor.memberId };
    case "system":
      return {
        actor_kind: "system",
        actor_service: parseServiceName(actor.serviceName),
      };
  }
}

File: actor.go
package telemetry

import "fmt"

type ActorKind string

type ServiceName string

const (
	ServiceBackend           ServiceName = "geul-backend"
	ServiceWeb               ServiceName = "geul-web"
	ServiceIdentity          ServiceName = "geul-identity"
	ServiceEditorCollab      ServiceName = "geul-collab"
	ServiceCDN               ServiceName = "geul-cdn"
	ServiceOG                ServiceName = "geul-og"
	ServiceAssetOptimizer    ServiceName = "geul-asset-optimizer"
	ServiceTranscoder        ServiceName = "geul-transcoder"
	ServiceWaveformProcessor ServiceName = "geul-waveform-processor"
)

var canonicalServiceNames = map[ServiceName]struct{}{
	ServiceBackend:           {},
	ServiceWeb:               {},
	ServiceIdentity:          {},
	ServiceEditorCollab:      {},
	ServiceCDN:               {},
	ServiceOG:                {},
	ServiceAssetOptimizer:    {},
	ServiceTranscoder:        {},
	ServiceWaveformProcessor: {},
}

func ParseServiceName(value string) (ServiceName, error) {
	serviceName := ServiceName(value)
	if _, ok := canonicalServiceNames[serviceName]; !ok {
		return "", fmt.Errorf("unknown canonical service name %q", value)
	}
	return serviceName, nil
}

func (serviceName ServiceName) String() string {
	return string(serviceName)
}

func (serviceName ServiceName) Instrumentation(component string) string {
	return string(serviceName) + "/" + component
}

const (
	ActorKindAnonymous ActorKind = "anonymous"
	ActorKindMember    ActorKind = "member"
	ActorKindSystem    ActorKind = "system"
)

// Actor is the runtime subject associated with a request. Only the concrete
// actor types in this package can satisfy the interface.
type Actor interface {
	Kind() ActorKind
	actor() ActorKind
}

type AnonymousActor struct{}

func (AnonymousActor) Kind() ActorKind  { return ActorKindAnonymous }
func (AnonymousActor) actor() ActorKind { return ActorKindAnonymous }

type MemberActor struct {
	SessionID  string
	IdentityID string
	MemberID   string
}

func (MemberActor) Kind() ActorKind  { return ActorKindMember }
func (MemberActor) actor() ActorKind { return ActorKindMember }

type SystemActor struct {
	ServiceName ServiceName
}

func (SystemActor) Kind() ActorKind  { return ActorKindSystem }
func (SystemActor) actor() ActorKind { return ActorKindSystem }

// RecordActor is the safe actor projection written to logs and audit records.
// Session and identity identifiers deliberately have no wire fields here.
type RecordActor struct {
	Kind     ActorKind `json:"actor_kind"`
	MemberID string    `json:"actor_member_id,omitempty"`
	Service  string    `json:"actor_service,omitempty"`
}

func ActorForRecord(actor Actor) (RecordActor, error) {
	if actor == nil {
		actor = AnonymousActor{}
	}
	switch typed := actor.(type) {
	case AnonymousActor:
		return RecordActor{Kind: ActorKindAnonymous}, nil
	case MemberActor:
		if typed.MemberID == "" {
			return RecordActor{}, fmt.Errorf("member actor requires member ID")
		}
		return RecordActor{Kind: ActorKindMember, MemberID: typed.MemberID}, nil
	case SystemActor:
		serviceName, err := ParseServiceName(string(typed.ServiceName))
		if err != nil {
			return RecordActor{}, err
		}
		return RecordActor{Kind: ActorKindSystem, Service: string(serviceName)}, nil
	default:
		return RecordActor{}, fmt.Errorf("unsupported actor type %T", actor)
	}
}

func (actor RecordActor) Validate() error {
	switch actor.Kind {
	case ActorKindAnonymous:
		if actor.MemberID != "" || actor.Service != "" {
			return fmt.Errorf("anonymous actor cannot contain member or service identity")
		}
	case ActorKindMember:
		if actor.MemberID == "" || actor.Service != "" {
			return fmt.Errorf("member actor requires only actor_member_id")
		}
	case ActorKindSystem:
		if actor.MemberID != "" {
			return fmt.Errorf("system actor requires only actor_service")
		}
		if _, err := ParseServiceName(actor.Service); err != nil {
			return err
		}
	default:
		return fmt.Errorf("invalid actor kind %q", actor.Kind)
	}
	return nil
}
